### ISSUE TITLE
[Question] Cross-cluster Federation for ephemeral clusters

### ISSUE THREAD

**Issue by whitleykeith** [thanos-io/thanos#4227]
I'm a performance engineer for OpenShift and I'm looking into Thanos as a solution to a problem we have with metrics reporting in ephemeral clusters. 

As a bit of context, my team routinely deploys ephemeral openshift clusters to do a number of different performance tests and then analyze the results to determine any regressions we see in upcoming releases. We frequently have to do manual snapshotting of grafana dashboards/manual tarballing of prom data in order to save all of these metrics which is both time-consuming and cumbersome. 

We've tried to use `remote_write` to pipe this data to an InfluxDB instance, but the amount of metrics would quickly overwhelm the system (we frequently test at 250/500/1k node scales and many times in parallel).

Ideally, I would like to leverage Thanos to have all of our ephemeral clusters backup their metrics into some store like S3 via a sidecar, and then have a Thanos querier instance deployed in our long-running infra cluster that can read all of these metrics so that we can have a one-stop shop for all of our performance metrics. 

Looking into it, the only two solutions I see for cross-cluster federation in Thanos is:

1. Have a "Observer" Thanos Querier instance that talks to the "Observee" clusters Thanos. This doesn't work for us because the Observee clusters are ephemeral and thus their Thanos Queriers are as well. [relevant doc](https://particule.io/en/blog/thanos-monitoring/)
2. Have each Prometheus instance remote_write to a centralized Thanos Receiver. I'm hesitant about this given the past experiences we've had with remote_write although this would probably be more performant with Thanos than with InfluxDB. [relevant doc](https://www.openshift.com/blog/federated-prometheus-with-thanos-receive)

What I would *really* like to see is something like this:

* each ephemeral cluster backs up to S3 via it's own Thanos sidecar
* a centralized thanos querier/store gateway that can hit all the buckets without having to reconfigure the gateway on each new cluster (i.e. give the store gateway an s3 bucket prefix that can map to many backups)

I'm not sure exactly how that would work so I wanted to see if any of the maintainers could give me some insight the feasibility of this setup?

**Comment by wiardvanrij** [comment]
Hi welcome! 

Cool to read your use-case. If any, feel free to hop into our community meetings if you ever feel to discuss this more in-depth via non-text :) 

Let me just start with that Thanos should be seen as a set of tools that you can leverage in quite some ways. Components can be used (or not), chained, ignored, etc. So there is a lot of flexibility. 

> Have a "Observer" Thanos Querier instance that talks to the "Observee" clusters Thanos. This doesn't work for us because the Observee clusters are ephemeral and thus their Thanos Queriers are as well.

That's actually not a problem. In fact, it's not even required to have a Thanos Querier in the observed clusters. It might be a bit easier but definitely not a requirement. 

> Have each Prometheus instance remote_write to a centralized Thanos Receiver. I'm hesitant about this given the past experiences we've had with remote_write although this would probably be more performant with Thanos than with InfluxDB.

I'm obviously biased here and I've used InfluxDB before. There is a reason I prefer Prometheus & Thanos. But regardless of my opinion, if we just look at more of the 'facts' - We can actually scale indefinitely. On every component. That's also what larger vendors that provide a 'PaaS/SaaS' are using under the hood (I.e. Cortex, which contains certain Thanos parts and visa-versa). So I'm actually convinced that performance would not become a problem. If it for some reason does, we are happy to hear about that and try to find a solution for it. 

It's just a matter of figuring out what would be the ideal setup in your use case. Honestly, there is no golden rule for this. Yet I can sketch a few examples:

![image](https://user-images.githubusercontent.com/5786097/118049158-9fe80900-b37d-11eb-8e5d-3f195c6da659.png)

![image](https://user-images.githubusercontent.com/5786097/118049186-aa0a0780-b37d-11eb-97e1-87d729ce924f.png)

![image](https://user-images.githubusercontent.com/5786097/118049197-afffe880-b37d-11eb-8010-a7022039bd2d.png)

![image](https://user-images.githubusercontent.com/5786097/118049215-b8f0ba00-b37d-11eb-990a-13533865cf0d.png)

So there are a few pro's and con's for each situation. There is also much to configure, think about scalability by sharding. Think about autoscaling (see for example this: https://github.com/observatorium/thanos-receive-controller) 
These are somewhat covered in the docs but also require some extra thoughts and information about what setup & requirements are used.


> each ephemeral cluster backs up to S3 via it's own Thanos sidecar 

If we use a sidecar for this, you'll have to think about the fact that this data then gets shipped every 2 hours (when a block is made). IF this is something you can incorporate in your ephemeral, this would work just fine. With incorporate I mean that you only destroy the cluster +2 hours after you are done. This is to be safe that all data is actually uploaded

> a centralized thanos querier/store gateway that can hit all the buckets without having to reconfigure the gateway on each new cluster

I hope my sketches give the insights that this is doable. It also does not matter if you use 1 bucket or 100 buckets. 

As last, something I did not draw, is basically using some form of setup with a certain mesh (i.e. cilium) so that a prometheus could discover (and therefore monitor) cross-cluster services. Yet I think you would be better of with something else such as remote_write as this is easier scalable than such setup IMO.

**Comment by whitleykeith** [comment]
Hey thanks for the response! This helped us look at the different options so we definitely appreciate the different designs.

For the remote_write option, how does the data from the receivers get uploaded to S3? Does the receiver behave similarly to the sidecar in that regard?

**Comment by onprem** [comment]
> For the remote_write option, how does the data from the receivers get uploaded to S3? Does the receiver behave similarly to the sidecar in that regard?

Yes, Receivers have Prometheus TSDBs running internally which produces a block every 2 hours by default, which then gets uploaded to the S3 bucket.

**Comment by whitleykeith** [comment]
Do those blocks get retained in memory and on any disk for the receiver? I can see that the receiver writes it to disk, but my receiver seems to never actually free memory until after the retention period.

**Comment by onprem** [comment]
> Do those blocks get retained in memory and on any disk for the receiver?

Yes, they are written as normal TSDB blocks to the disk, you can customize to location using `--tsdb.path` flag.

> I can see that the receiver writes it to disk, but my receiver seems to never actually free memory until after the retention period.

Receiver is internally using the Prometheus TSDB implementation so it shouldn't be doing anything special. I remember Prometheus uses mmap for fast access to the data and the memory usage might be related to that.

> I also see that I'm getting a lot of "empty time-series from client" logs.

Those are because Receiver currently doesn't ingests metric metadata using Remote Write but Prometheus sends them by default. The metadata remote write requests does not contain any timeseries which leads to the error. It's pretty harmless so you can ignore that for now, or you can explicitly disable sending metadata in remote write from your Prometheus instances,

**Comment by whitleykeith** [comment]
> Receiver is internally using the Prometheus TSDB implementation so it shouldn't be doing anything special. I remember Prometheus uses mmap for fast access to the data and the memory usage might be related to that.

So I stopped remote_write for all clusters except for the one Thanos is hosted in (relatively small cluster with little activity) and even overnight the memory footprint was still a massive 250gb/pod. The disk usage on each pod was less than 10gb each which doesn't seem to make sense to me? Moreover, once I killed the pod and let them come back up it's only using around 1gb/memory each. Maybe this is worth of a separate issue but it seems like there may be a bug with the receiver
